Move generic arg/param validation to `create_substs_for_generic_args` to resolve various const generics issues

varkor opened the pull request:
This changes some diagnostics, but I think they're around as helpful as the previous ones, and occur infrequently regardless.

Fixes https://github.com/rust-lang/rust/issues/68257.
Fixes https://github.com/rust-lang/rust/issues/68398.

r? @eddyb

rust-highfive commented:
Some changes occurred in diagnostic error codes

cc @GuillaumeGomez

Centril commented on src/librustc_typeck/astconv.rs:
```diff
@@ -562,28 +574,74 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                                        arg_kind(arg),
+                                        param_kind,
+                                    )
+                                    .emit();
```
Can we move this out (a lot of rightward drift here + diagnostics code in happy path)?

Centril replied:
Also, `structural_impls` has:
```rust
impl fmt::Debug for ty::GenericParamDef {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let type_name = match self.kind {
            ty::GenericParamDefKind::Lifetime => "Lifetime",
            ty::GenericParamDefKind::Type { .. } => "Type",
            ty::GenericParamDefKind::Const => "Const",
        };
        write!(f, "{}({}, {:?}, {})", type_name, self.name, self.def_id, self.index)
    }
}
```
Consider sharing the `match` code.

Centril commented on src/librustc_typeck/astconv.rs:
```diff
@@ -562,28 +574,74 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                                arg_kind(provided),
+                                kind,
+                            )
+                            .emit();
```
Can we move this out too (+ `arg_kind`)?

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=19033&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/19033/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/19033/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-01-21T23:11:56.1184654Z ========================== Starting Command Output ===========================
2020-01-21T23:11:56.1187464Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/a3ac8a1c-1c0b-45ae-a72c-3169516ffc07.sh
2020-01-21T23:11:56.1187627Z 
2020-01-21T23:11:56.1191330Z ##[section]Finishing: Disable git automatic line ending conversion
2020-01-21T23:11:56.1196322Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/68434/merge to s
2020-01-21T23:11:56.1197901Z Task         : Get sources
2020-01-21T23:11:56.1197930Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-01-21T23:11:56.1197966Z Version      : 1.0.0
2020-01-21T23:11:56.1197994Z Author       : Microsoft
---
2020-01-21T23:11:56.9168959Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-01-21T23:11:56.9257079Z ##[command]git config gc.auto 0
2020-01-21T23:11:56.9310430Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-01-21T23:11:56.9357922Z ##[command]git config --get-all http.proxy
2020-01-21T23:11:56.9507255Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/68434/merge:refs/remotes/pull/68434/merge
---
2020-01-21T23:17:12.4893650Z    Compiling serde_json v1.0.40
2020-01-21T23:17:13.9663480Z    Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
2020-01-21T23:17:22.9877832Z     Finished release [optimized] target(s) in 1m 16s
2020-01-21T23:17:22.9976715Z tidy check
2020-01-21T23:17:23.5154947Z tidy error: /checkout/src/librustc_error_codes/error_codes/E0747.md:1: line longer than 80 chars
2020-01-21T23:17:23.5155052Z tidy error: /checkout/src/librustc_error_codes/error_codes/E0747.md:9: line longer than 80 chars
2020-01-21T23:17:25.5252912Z Found 488 error codes
2020-01-21T23:17:25.5253011Z Found 0 error codes with no tests
2020-01-21T23:17:25.5253057Z Done!
2020-01-21T23:17:25.5253136Z some tidy checks failed
2020-01-21T23:17:25.5253136Z some tidy checks failed
2020-01-21T23:17:25.5257118Z 
2020-01-21T23:17:25.5257184Z 
2020-01-21T23:17:25.5258097Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor"
2020-01-21T23:17:25.5258258Z 
2020-01-21T23:17:25.5258283Z 
2020-01-21T23:17:25.5267254Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
2020-01-21T23:17:25.5267319Z Build completed unsuccessfully in 0:01:26
2020-01-21T23:17:25.5267319Z Build completed unsuccessfully in 0:01:26
2020-01-21T23:17:25.5313415Z == clock drift check ==
2020-01-21T23:17:25.5323311Z   local time: Tue Jan 21 23:17:25 UTC 2020
2020-01-21T23:17:25.6946370Z   network time: Tue, 21 Jan 2020 23:17:25 GMT
2020-01-21T23:17:25.6956027Z == end clock drift check ==
2020-01-21T23:17:26.3853952Z 
2020-01-21T23:17:26.3956496Z ##[error]Bash exited with code '1'.
2020-01-21T23:17:26.3971596Z ##[section]Finishing: Run build
2020-01-21T23:17:26.3987887Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/68434/merge to s
2020-01-21T23:17:26.3989863Z Task         : Get sources
2020-01-21T23:17:26.3989959Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-01-21T23:17:26.3990129Z Version      : 1.0.0
2020-01-21T23:17:26.3990166Z Author       : Microsoft
2020-01-21T23:17:26.3990166Z Author       : Microsoft
2020-01-21T23:17:26.3990225Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2020-01-21T23:17:26.3990270Z ==============================================================================
2020-01-21T23:17:26.7706649Z Cleaning any cached credential from repository: rust-lang/rust (GitHub)
2020-01-21T23:17:26.7748375Z ##[section]Finishing: Checkout rust-lang/rust@refs/pull/68434/merge to s
2020-01-21T23:17:26.7857429Z Cleaning up task key
2020-01-21T23:17:26.7858475Z Start cleaning up orphan processes.
2020-01-21T23:17:26.7955913Z Terminate orphan process: pid (4037) (python)
2020-01-21T23:17:26.8137367Z ##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

eddyb commented on src/librustc/ty/structural_impls.rs:
```diff
@@ -19,12 +19,7 @@ use syntax::ast;
-            ty::GenericParamDefKind::Const => "Const",
-        };
-        write!(f, "{}({}, {:?}, {})", type_name, self.name, self.def_id, self.index)
+        write!(f, "{}({}, {:?}, {})", self.kind.descr(), self.name, self.def_id, self.index)
```
The casing is different though - consider deriving `Debug` instead of this custom `impl`.

eddyb commented on src/librustc_typeck/astconv.rs:
```diff
@@ -480,6 +496,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         parent_substs: &[subst::GenericArg<'tcx>],
         has_self: bool,
         self_ty: Option<Ty<'tcx>>,
+        arg_count_mismatch: bool,
```
Hmm, maybe you can replace this with a `Result<(), GenericArgCountMismatch>`?
Especially if it's the same `bool` returned by some functions.

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=19138&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/19138/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/19138/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-01-23T00:42:09.2818335Z ========================== Starting Command Output ===========================
2020-01-23T00:42:09.2821350Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/d3c7e16e-df35-4350-8282-6aee5ef18887.sh
2020-01-23T00:42:09.2821540Z 
2020-01-23T00:42:09.2867194Z ##[section]Finishing: Disable git automatic line ending conversion
2020-01-23T00:42:09.2873297Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/68434/merge to s
2020-01-23T00:42:09.2875351Z Task         : Get sources
2020-01-23T00:42:09.2875394Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-01-23T00:42:09.2875427Z Version      : 1.0.0
2020-01-23T00:42:09.2875458Z Author       : Microsoft
---
2020-01-23T00:42:10.2199341Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-01-23T00:42:10.2298342Z ##[command]git config gc.auto 0
2020-01-23T00:42:10.2370098Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-01-23T00:42:10.2423655Z ##[command]git config --get-all http.proxy
2020-01-23T00:42:10.2618275Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/68434/merge:refs/remotes/pull/68434/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

eddyb commented on src/librustc_typeck/astconv.rs:
```diff
@@ -337,112 +341,128 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             && (!infer_lifetimes || arg_counts.lifetimes > param_counts.lifetimes)
         {
-            check_kind_count(
+            arg_count_mismatch |= check_kind_count(
```
AFAIK you can have a `Result` variable and combine it with `x = x.and(...)`.

varkor replied:
I don't think the code becomes more readable using a `Result` instead of `bool` here.

eddyb commented on src/librustc_typeck/astconv.rs:
```diff
@@ -337,112 +341,128 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                // For kinds without defaults (e.g.., lifetimes), `required == permitted`.
+                // For other kinds (i.e., types), `permitted` may be greater than `required`.
+                if required <= provided && provided <= permitted {
+                    return false;
```
At the very least, I'd like to see this closure return `Result` too.
I can't easily tell if `false` here is success or failure, but `Ok(())` would be clear.

eddyb commented on src/librustc_typeck/astconv.rs:
```diff
@@ -314,9 +318,10 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
 
         // Prohibit explicit lifetime arguments if late-bound lifetime parameters are present.
-        let mut reported_late_bound_region_err = None;
+        let mut reported_late_bound_region_err = false;
```
This could probably also avoid being `bool`.

eddyb commented on src/librustc_typeck/astconv.rs:
```diff
@@ -327,7 +332,6 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                     let mut err = tcx.sess.struct_span_err(span, msg);
                     err.span_note(span_late, note);
                     err.emit();
-                    reported_late_bound_region_err = Some(true);
```
`reported_late_bound_region_err.unwrap_or(false)` below would suggest that only this branch should have `reported_late_bound_region_err = true;`, not the lint one below?

varkor replied:
I can change the variable name. I thought it was reasonable to include emission of a lint as "reported", though, so the name still seemed appropriate.

eddyb replied:
Doesn't that change the behavior later on (potentially preventing real errors from being emitted)?

The important thing, I thought, is whether there was an actual error that would stop compilation (maybe we should use `ErrorReported` to indicate that?).

varkor replied:
It's possible there are interactions with `impl Trait` that aren't being tested at all. I'll check this.

eddyb replied:
I more or less really want to keep the preexisting behavior, if the tests pass either way.

estebank commented on src/librustc_typeck/astconv.rs:
```diff
@@ -337,112 +341,128 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+            arg.descr(),
+            kind,
+        )
+        .emit();
```
No reordering suggestion for this case?

varkor replied:
I'd rather leave this PR for the bug-fixes and follow up with the diagnostics improvements. That's a good follow up, though.

varkor commented:
r? @nikomatsakis for a second opinion (https://github.com/rust-lang/rust/pull/68434#pullrequestreview-346257567)

eddyb commented on src/librustc_typeck/astconv.rs:
```diff
@@ -321,18 +322,19 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                    explicit_lifetimes = Err(GenericArgCountMismatch(Some(ErrorReported)));
                     let mut err = tcx.sess.struct_span_err(span, msg);
                     err.span_note(span_late, note);
                     err.emit();
```
@Centril @estebank Can we make it so you have to emit an error to get an `ErrorReported` token, and maybe make it unconstructible except by the diagnostic reporting machinery?

nikomatsakis replied:
Oh, that'd be cute. :)

varkor replied:
Are we tracking this anywhere?

eddyb replied:
Don't think so.

varkor replied:
Created https://github.com/rust-lang/rust/issues/69426.

yodaldevoid commented:
Is this PR waiting on anything specific?

varkor commented:
@yodaldevoid: a second reviewer's opinion on the PR.

bors commented:
:umbrella: The latest upstream changes (presumably #68893) made this pull request unmergeable. Please resolve the merge conflicts.

estebank commented on src/librustc_error_codes/error_codes/E0747.md:
````diff
@@ -0,0 +1,11 @@
+
+type X = S<(), 'static>; // error: the type argument is provided before the
+                         // lifetime argument
+```
````
Add the fixed example.

estebank commented on src/test/ui/const-generics/const-param-after-const-literal-arg.rs:
```diff
@@ -0,0 +1,10 @@
+// check-pass
+
+#![feature(const_generics)]
+//~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash
```
Can't we just add `#![allow(incomplete_features)]` instead?

estebank commented on src/test/ui/const-generics/const-param-before-other-params.rs:
```diff
@@ -1,4 +1,5 @@
 #![feature(const_generics)]
+//~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash
```
Same here.

estebank commented on src/test/ui/suggestions/suggest-move-types.stderr:
```diff
@@ -74,29 +74,30 @@ LL | struct Dl<'a, 'b, 'c, T, U, V, M: ThreeWithLifetime<T, 'a, A=(), B=(), U, '
 
-error: lifetime arguments must be declared prior to type arguments
-  --> $DIR/suggest-move-types.rs:34:46
+error[E0747]: type provided when a lifetime was expected
```
I'm convinced we should be suggesting the correct order in all of these cases.

varkor replied:
The reason I changed this is because there isn't really a restriction on the order of generic arguments: the restriction is on generic parameters, but generic arguments must match the order of generic parameters. The fact that lifetime arguments must precede type arguments is a side-effect of the parameter order. I'll try tweaking the wording to improve this. As @estebank says, ideally we'd suggest the correct order, but I'd like to get this merged sooner rather than later, as I don't know how much time I'll have for more elaborate changes now.

nikomatsakis replied:
OK, I see. I still find the message more confusing than the old one, but perhaps I now better understand what this PR is aiming for.

varkor commented on src/test/ui/generic/generic-arg-mismatch-recover.rs:
```diff
@@ -4,7 +4,6 @@ struct Bar<'a>(&'a ());
 
 fn main() {
     Foo::<'static, 'static, ()>(&0); //~ ERROR wrong number of lifetime arguments
```
Previously if we encountered an incorrect generic argument, we would just try to skip it and continue going: https://github.com/rust-lang/rust/pull/68434/files#diff-1b4d50b82a87f1031cd5f99b64123de7L578. Now we just stop checking the remaining arguments, because there's a high likelihood the remaining arguments are invalid and we'd just spit out lots of errors that aren't particularly helpful.

varkor commented:
@nikomatsakis: I've addressed the code comments, and also added a note to the error messages making them as descriptive as the previous ones. I'll open an issue for suggesting the correct order of all provided arguments, though I'd rather split it out into a separate PR.

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=21515&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21515/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21515/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-02-22T02:36:28.8377938Z ========================== Starting Command Output ===========================
2020-02-22T02:36:28.8380945Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/329cb1fd-4786-467a-ad53-100ffddcecfd.sh
2020-02-22T02:36:28.8381166Z 
2020-02-22T02:36:28.8385052Z ##[section]Finishing: Disable git automatic line ending conversion
2020-02-22T02:36:28.8406108Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/68434/merge to s
2020-02-22T02:36:28.8410059Z Task         : Get sources
2020-02-22T02:36:28.8410307Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-02-22T02:36:28.8411269Z Version      : 1.0.0
2020-02-22T02:36:28.8411447Z Author       : Microsoft
---
2020-02-22T02:36:30.4658369Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-02-22T02:36:30.4809836Z ##[command]git config gc.auto 0
2020-02-22T02:36:30.4851864Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-02-22T02:36:30.4882765Z ##[command]git config --get-all http.proxy
2020-02-22T02:36:30.4968550Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/68434/merge:refs/remotes/pull/68434/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis commented:
@bors r+

thanks, @varkor

bors commented:
:pushpin: Commit bead79ebc6d0f605f42a0f8315ce9f5fe5764b99 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis bead79ebc6d0f605f42a0f8315ce9f5fe5764b99 -->
<!-- homu: {"type":"Approved","sha":"bead79ebc6d0f605f42a0f8315ce9f5fe5764b99","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit bead79ebc6d0f605f42a0f8315ce9f5fe5764b99 with merge 6d69caba110c0c2fb90180df1cbc8be5033b91d4...
<!-- homu: {"type":"BuildStarted","head_sha":"bead79ebc6d0f605f42a0f8315ce9f5fe5764b99","merge_sha":"6d69caba110c0c2fb90180df1cbc8be5033b91d4"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21910)
Approved by: nikomatsakis
Pushing 6d69caba110c0c2fb90180df1cbc8be5033b91d4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21910"},"merge_sha":"6d69caba110c0c2fb90180df1cbc8be5033b91d4"} -->

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#68434!

Tested on commit rust-lang/rust@6d69caba110c0c2fb90180df1cbc8be5033b91d4.
Direct link to PR: <https://github.com/rust-lang/rust/pull/68434>

🎉 book on windows: test-fail → test-pass (cc @carols10cents @steveklabnik).
